File: 1127-last-stone-weight/last-stone-weight.ts
function lastStoneWeight(stones: number[]): number {
  if (stones.length === 1) return stones[0] ?? 0;

  stones.sort((a, b) => b - a);

  while (stones.length > 1) {
    const stone1 = stones.shift()!;
    const stone2 = stones.shift()!;

    if (stone1 !== stone2) {
      const newStone = stone1 - stone2;
      const index = stones.findIndex((s) => s < newStone);
      stones.splice(index >= 0 ? index : stones.length, 0, newStone);
    }
  }

  return stones.length === 0 ? 0 : stones[0];
}

File: 1127-last-stone-weight/last_stone_weight.rb
def last_stone_weight(stones)
  return stones[0] if stones.length == 1

  stones.sort!.reverse!

  while stones.length > 1
    stone1 = stones.shift
    stone2 = stones.shift

    if stone1 != stone2
      new_stone = stone1 - stone2
      index = stones.index { |s| s < new_stone } || stones.length
      stones.insert(index, new_stone)
    end
  end

  stones.empty? ? 0 : stones[0]
end